docs/src/design/proposals/APIRedesign_PeteRichards.md

```diff
@@ -0,0 +1,251 @@
+- [[1] Footnote: The angular debacle](#1-footnote-the-angular-debacle)
+  - ["Do or do not, there is no try"](#do-or-do-not-there-is-no-try)
+  - [A lack of commitment](#a-lack-of-commitment)
+  - [Commitment is good!](#commitment-is-good)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+
+# Reducing interface depth (the bundle.json version)
+
+## Imperitive component registries 
+
+Transition component registries to javascript, get rid of bundle.json and bundles.json.  Prescribe a method for application configuration, but allow flexibility in how application configuration is defined.
+
+Register components in an imperitive fashion, see angularApp.factory, angularApp.controller, etc.  Alternatively, implement our own application object with new registries and it's own form of registering objects.
+
+## Get rid of "extension category" concept.
+
+The concept of an "extension category" is itself an extraneous concept-- an extra layer of interface depth, an extra thing to learn before you can say "hello world".  Extension points should be clearly supported and documented with whatever interfaces make sense.  Developers who wish to add something that is conceptually equivalent to an extension category can do so directly, in the manner that suites their needs, without us forcing a common method on them.
+
+## Reduce number and depth of extension points
+
+Clearly specify supported extension points (e.g. persistence, model providers, telemetry providers, routes, time systems), but don't claim that the system has a clear and perfect repeatable solution for unknown extension types.  New extension categories can be implemented in whatever way makes sense, without prescribing "the one and only system for managing extensions".
+
+The underlying problem here is we are predicting needs for extension points where none exist-- if we try and design the extension system before we know how it is used, we design the wrong thing and have to rewrite it later.
+
+## Composite services should not be the default
+
+Understanding composite services, and describing services as composite services can confuse developers.  Aggregators are implemented once and forgotten, while decorators tend to be hacky, brittle solutions that are generally needed to avoid circular imports.  While composite services are a useful construct, it reduces interface depth to implement them as registries + typed providers.
+
+You can write a provider (provides "thing x" for "inputs y") with a simple interface.  A provider has two or more methods:
+* a method which takes "inputs y" and returns True if it knows how to provide "thing x", false otherwise.
+* one or more methods which provide "thing x" for objects of "inputs y".  
+
+Actually checking whether a provider can respond to a request before asking it to do work allows for faster failure and clearer errors when no providers match the request.
```

Comment:
Decorators are available in Angular (`$provide.decorate` has been around for a while, more recently `decorate` was hoisted up to the module level) so we can factor that out easily. Services can be decorated by default.

That leaves service compositing. Agree that we should support this in a way that doesn't require developers to understand that detail, but also think we should make it easy to support this by default. Treating these like registries (from an interface perspective) and providing a reusable `CompositeServiceProvider` seems like it should be sufficient.

Having the "providers" expose a different interface from the composite service itself is debatable. In some cases this improves the clarity of interfaces, but in all cases it increases the number of interfaces. In general I prefer the latter, but we could consider this trade case-by-case as well.